File: icamera/src/main/java/me/shouheng/icamera/preview/impl/TexturePreview.java
package me.shouheng.icamera.preview.impl;

import android.content.Context;
import android.graphics.SurfaceTexture;
import android.support.annotation.Nullable;
import android.view.Surface;
import android.view.SurfaceHolder;
import android.view.TextureView;
import android.view.View;
import android.view.ViewGroup;

import me.shouheng.icamera.enums.PreviewViewType;

/**
 * Camera preview implementation based on {@link TextureView}.
 *
 * @author WngShhng (shouheng2015@gmail.com)
 * @version 2019/4/13 22:54
 */
public class TexturePreview extends BaseCameraPreview {

    private SurfaceTexture surfaceTexture;
    private TextureView textureView;

    public TexturePreview(Context context, ViewGroup parent) {
        super(context, parent);
        textureView = new TextureView(context);
        textureView.setLayoutParams(new ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));
        parent.addView(textureView);
        textureView.setSurfaceTextureListener(new TextureView.SurfaceTextureListener() {
            /**
             * Invoked when a {@link TextureView}'s SurfaceTexture is ready for use.
             *
             * @param surface The surface returned by
             *                {@link android.view.TextureView#getSurfaceTexture()}
             * @param width   The width of the surface
             * @param height  The height of the surface
             */
            @Override
            public void onSurfaceTextureAvailable(SurfaceTexture surface, int width, int height) {
                updateSurfaceTexture(surface, width, height);
                notifyPreviewAvailable();
            }

            /**
             * Invoked when the {@link SurfaceTexture}'s buffers size changed.
             *
             * @param surface The surface returned by
             *                {@link android.view.TextureView#getSurfaceTexture()}
             * @param width   The new width of the surface
             * @param height  The new height of the surface
             */
            @Override
            public void onSurfaceTextureSizeChanged(SurfaceTexture surface, int width, int height) {
                updateSurfaceTexture(surface, width, height);
                notifyPreviewAvailable();
            }

            /**
             * Invoked when the specified {@link SurfaceTexture} is about to be destroyed.
             * If returns true, no rendering should happen inside the surface texture after this method
             * is invoked. If returns false, the client needs to call {@link SurfaceTexture#release()}.
             * Most applications should return true.
             *
             * @param surface The surface about to be destroyed
             */
            @Override
            public boolean onSurfaceTextureDestroyed(SurfaceTexture surface) {
                updateSurfaceTexture(surface, 0, 0);
                return true;
            }

            @Override
            public void onSurfaceTextureUpdated(SurfaceTexture surface) {
                /* noop */
            }
        });
    }

    @Override
    public Surface getSurface() {
        return new Surface(surfaceTexture);
    }

    @Override
    @PreviewViewType
    public int getPreviewType() {
        return PreviewViewType.TEXTURE_VIEW;
    }

    @Nullable
    @Override
    public SurfaceHolder getSurfaceHolder() {
        return null;
    }

    @Nullable
    @Override
    public SurfaceTexture getSurfaceTexture() {
        return surfaceTexture;
    }

    @Override
    public View getView() {
        return textureView;
    }

    /*-----------------------------------------inner methods---------------------------------------------*/

    private void updateSurfaceTexture(SurfaceTexture surfaceTexture, int width, int height) {
        this.surfaceTexture = surfaceTexture;
        setSize(width, height);
    }
}

File: icamera/src/main/java/me/shouheng/icamera/preview/impl/TexturePreview.kt
package me.shouheng.icamera.preview.impl

import android.content.Context
import android.graphics.SurfaceTexture
import android.view.*
import android.view.TextureView.SurfaceTextureListener
import me.shouheng.icamera.enums.PreviewViewType

/**
 * Camera preview implementation based on [TextureView].
 *
 * @author WngShhng (shouheng2015@gmail.com)
 * @version 2019/4/13 22:54
 */
class TexturePreview(context: Context, parent: ViewGroup) : BaseCameraPreview() {

    override var surfaceTexture: SurfaceTexture? = null
        private set

    private val textureView: TextureView = TextureView(context)

    override val surface: Surface
        get() = Surface(surfaceTexture)

    @get:PreviewViewType
    override val previewType: Int
        get() = PreviewViewType.TEXTURE_VIEW

    override var surfaceHolder: SurfaceHolder? = null

    override val view: View
        get() = textureView

    /*-----------------------------------------inner methods---------------------------------------------*/
    private fun updateSurfaceTexture(surfaceTexture: SurfaceTexture, width: Int, height: Int) {
        this.surfaceTexture = surfaceTexture
        setSize(width, height)
    }

    init {
        textureView.layoutParams = ViewGroup.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT
        )
        parent.addView(textureView)
        textureView.surfaceTextureListener = object : SurfaceTextureListener {
            /**
             * Invoked when a [TextureView]'s SurfaceTexture is ready for use.
             *
             * @param surface The surface returned by
             * [android.view.TextureView.getSurfaceTexture]
             * @param width   The width of the surface
             * @param height  The height of the surface
             */
            override fun onSurfaceTextureAvailable(surface: SurfaceTexture, width: Int, height: Int) {
                updateSurfaceTexture(surface, width, height)
                notifyPreviewAvailable()
            }

            /**
             * Invoked when the [SurfaceTexture]'s buffers size changed.
             *
             * @param surface The surface returned by
             * [android.view.TextureView.getSurfaceTexture]
             * @param width   The new width of the surface
             * @param height  The new height of the surface
             */
            override fun onSurfaceTextureSizeChanged(surface: SurfaceTexture, width: Int, height: Int) {
                updateSurfaceTexture(surface, width, height)
                notifyPreviewAvailable()
            }

            /**
             * Invoked when the specified [SurfaceTexture] is about to be destroyed.
             * If returns true, no rendering should happen inside the surface texture after this method
             * is invoked. If returns false, the client needs to call [SurfaceTexture.release].
             * Most applications should return true.
             *
             * @param surface The surface about to be destroyed
             */
            override fun onSurfaceTextureDestroyed(surface: SurfaceTexture): Boolean {
                updateSurfaceTexture(surface, 0, 0)
                return true
            }

            override fun onSurfaceTextureUpdated(surface: SurfaceTexture) {/* noop */ }
        }
    }
}